From dd32dd7a885eca6477ea8878780f1f5fdedf81e9 Mon Sep 17 00:00:00 2001
From: Evan Huus <eapache@gmail.com>
Date: Fri, 29 Mar 2013 22:20:04 +0000
Subject: [PATCH 5/7] Bump two guint16 to guint32 to prevent overflow when
 reassembling a large number of fragments, and add an
 extra bounds check.

Fixes
https://bugs.wireshark.org/bugzilla/show_bug.cgi?id=8540
https://bugs.wireshark.org/bugzilla/show_bug.cgi?id=8541

svn path=/trunk/; revision=48644
---
 epan/dissectors/packet-dcp-etsi.c |    6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/epan/dissectors/packet-dcp-etsi.c b/epan/dissectors/packet-dcp-etsi.c
index c646a08..a90407c 100644
--- a/epan/dissectors/packet-dcp-etsi.c
+++ b/epan/dissectors/packet-dcp-etsi.c
@@ -260,7 +260,7 @@ dissect_pft_fec_detailed(tvbuff_t * tvb, packet_info * pinfo, proto_tree * tree,
   fragment_data *fdx
 )
 {
-  guint16 decoded_size;
+  guint32 decoded_size;
   guint32 c_max;
   guint32 rx_min;
   tvbuff_t *new_tvb=NULL;
@@ -339,11 +339,11 @@ dissect_pft_fec_detailed(tvbuff_t * tvb, packet_info * pinfo, proto_tree * tree,
                                             NULL, tree);
     }
   }
-  if(new_tvb) {
+  if(new_tvb && tvb_length(new_tvb) > 0) {
     gboolean decoded = TRUE;
     tvbuff_t *dtvb = NULL;
     const guint8 *input = tvb_get_ptr(new_tvb, 0, -1);
-    guint16 reassembled_size = tvb_length(new_tvb);
+    guint32 reassembled_size = tvb_length(new_tvb);
     guint8 *deinterleaved = (guint8*) g_malloc (reassembled_size);
     guint8 *output = (guint8*) g_malloc (decoded_size);
     rs_deinterleave(input, deinterleaved, plen, fcount);
-- 
1.7.10.4

